oauth_dropins/meetup.py

```diff
@@ -0,0 +1,162 @@
+"""Meetup.com drop-in.
+
+API docs:
+https://www.meetup.com/meetup_api/
+"""
+import logging
+import urllib.parse
+
+from google.cloud import ndb
+from webob import exc
+
+from . import handlers
+from .models import BaseAuth
+from .webutil import appengine_config, appengine_info, util
+from .webutil.util import json_dumps, json_loads
+
+if appengine_info.DEBUG:
+    MEETUP_CLIENT_ID = util.read('meetup_client_id_local')
+    MEETUP_CLIENT_SECRET = util.read('meetup_client_secret_local')
+else:
+    MEETUP_CLIENT_ID = util.read('meetup_client_id')
+    MEETUP_CLIENT_SECRET = util.read('meetup_client_secret')
+
+GET_AUTH_CODE_URL = '&'.join((
+    'https://secure.meetup.com/oauth2/authorize'
+    'client_id=%{client_id}s',
+    'response_type=code',
+    'redirect_uri=%{redirect_uri}s',
+    'state=%(state)s',
+    'scope=rsvp'
+    ))
+
+GET_ACCESS_TOKEN_URL = '&'.join((
+    'https://secure.meetup.com/oauth2/access'
+    'grant_type=authorization_code',
+    'code=%(code)s',
+    'client_id=%(client_id)s',
+    'client_secret=%(client_secret)s',
+    'redirect_uri=%(redirect_uri)s',
+    ))
+
+GET_USER_INFO_URL = 'https://api.meetup.com/members/self/'
+
+
+class MeetupAuth(BaseAuth):
+    """An authenticated Meetup.com user.
+
+    Provides methods that return information about this user and make
+    OAuth-signed requests to Meetup's HTTP-based APIs. Stores OAuth credentials
+    in the datastore. See models.BaseAuth for usage details.
+
+    Implements urlopen() but not http() or api().
+    """
+    access_token_str = ndb.StringProperty(required=True)
+
+    def site_name(self):
+        return 'Meetup.com'
+
+    def user_display_name(self):
+        """Returns the Meetup.com user id.
+        """
+        return self.key.string_id()
+
+
+def access_token(self):
```

Comment:
this line through 68 need to be indented to be methods of `MeetupAuth` instead of top-level functions